File: 283/Solution.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var moveZeroes = function (nums) {
  if (!nums) return;

  /**
   * 看前面的那个python解法就好了
   * 这个代码是最初写的
   * 没用之前那个思路
   * 直接用的JS提供的 Array.prototype.splice方法
   * 这个方法能删除当前位置的元素 并且使后面的元素挨个排上来
   *
   * 这个代码的思路也是双指针
   * 不过right指针只用作边界
   * 首先right指针从右往左找到第一个不是0的数 这里就是遍历的边界
   * 然后left从左开始出发
   *    如果是0的话 则删除当前元素 并在数组末尾添加0
   *    注意删除元素的时候 left不用再动了 而right需要--
   */
  let left = 0,
    right = nums.length - 1;
  while (right >= 0 && nums[right] === 0) {
    right--;
  }

  while (left < right) {
    if (nums[left] === 0) {
      nums.splice(left, 1);
      nums.push(0);
      right--;
      continue;
    }
    left++;
  }
};

File: 283/Solution.py
from typing import List


class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        if not nums:
            return

        """
        首先说一下思路, 这道题并不难
        双指针 left 和 right
        right出去寻找不为0的数 并将其与left对应的值交换

        初始left和right均为0
        若第一个是0  比如 0 1 2 
            那么right就会找到1  随后0和1交换 left++ right++
        若第一个是非0数 比如 1 0 2
            那么right找到的第一个非零数就是自己  自然不会影响结果 然后left和right分别后移
        
        left左边都是按照原来相对顺序排列的非零数
        left-right中间都是0

        时间复杂度: O(n) 只需要一次遍历数组
        空间复杂度: O(1)
        """

        size = len(nums)
        left, right = 0, 0
        while right < size:
            if nums[right] != 0:
                nums[left], nums[right] = nums[right], nums[left]
                left += 1
            right += 1
